### ISSUE TITLE
Link Stats Not Loading on Self Hosting

### ISSUE THREAD

**Issue by berkyzl** [dubinc/dub#505]
Hello, thanks for the guide about self hosting. I followed all the steps and I was able to run dub and shorten links.
But when I click on link stats, the stats keep loading. After checking logs on vercel I saw few errors about /api/stats but didn't know how to fix it.

Endpoints effected:
/api/stats/device
/api/stats/timeseries
/api/stats/country
/api/stats/referer

I'm able to see clicks on the link list, but can't see it in the stats page.
![image](https://github.com/dubinc/dub/assets/143607443/30b98bde-61d9-4ab7-80b8-eb0f1c7159de)
![image](https://github.com/dubinc/dub/assets/143607443/e47bafdf-7a4b-44b0-97d9-53b3d926fcd8)

`⨯ TypeError: Value is not JSON serializable
    at serializeJavascriptValueToJSONString (node:internal/deps/undici/undici:1185:15)
    at Response.json (node:internal/deps/undici/undici:6728:55)
    at u.json (/var/task/apps/web/.next/server/chunks/4688.js:1:10017)
    at i.needNotExceededUsage (/var/task/apps/web/.next/server/chunks/6356.js:1:502)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /var/task/node_modules/.pnpm/next@14.0.4-canary.39_@babel+core@7.17.8_@opentelemetry+api@1.4.1_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:42380`
    
 `TypeError: Cannot read properties of undefined (reading '0')
    at /var/task/apps/web/.next/server/chunks/6356.js:1:9989
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async I (/var/task/apps/web/.next/server/chunks/6356.js:1:9832)
    at async i.needNotExceededUsage (/var/task/apps/web/.next/server/chunks/6356.js:1:431)
    at async /var/task/node_modules/.pnpm/next@14.0.4-canary.39_@babel+core@7.17.8_@opentelemetry+api@1.4.1_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-route.runtime.prod.js:6:42380`

**Comment by steven-tey** [comment]
Thanks a lot for reporting this! Are you running into this locally as well? If so, can you modify [this line](https://github.com/dubinc/dub/blob/main/apps/web/lib/stats.ts#L165) to log the error message that you're getting from the Tinybird API endpoints?

**Comment by berkyzl** [comment]
I haven't run the project locally yet and I'm not fully versed in Next.js as of now. However, I plan to thoroughly review the project and should be able to provide more detailed feedback after that. I'm also interested in contributing to the project and look forward to sharing my learnings as I progress. While I don't have experience with modifying the line you mentioned to log error messages from the Tinybird API endpoints, I believe I can be more helpful after I have delved deeper into the project.

**Comment by berkyzl** [comment]
Also I'm not able to login to admin.[domain]. İs it beacause DUB_PROJECT_ID is hardcoded in utils?
in [this line](https://github.com/dubinc/dub/blob/4a61df3e6b0138e6d0afaa230b89ac67c04d0bf3/apps/web/app/admin.dub.co/actions.ts#L18), but event after adding it as environment variable, I wasn't able to login. It keeps redirecting me to login page.

**Comment by steven-tey** [comment]
Okay, please dive into it and let me know what you find out! 

As for admin, you probably don't need it – it's mostly for internal use at Dub for removing abusive links 😄

**Comment by berkyzl** [comment]
Hello, I added some logs. Double checked my Tinybird API, `tb push` endpoints.
I'm not familiar with tinybird or pipes. Any suggestions?

```js
  return await fetch(url.toString(), {
    headers: {
      Authorization: `Bearer ${process.env.TINYBIRD_API_KEY}`,
    },
  })
  .then((res) => {
    console.log("Response before JSON conversion:", res); //Added this log
    return res.json();
  })
    .then(({ data }) => {
      console.log("Data received from Tinybird:", data); //Added this log
      if (endpoint === "clicks") {
        try {
          const clicks = data[0]["count()"];
          return clicks || "0";
        } catch (e) {
          console.log(e);
        }
      }
      return data;
    });
```

Logs:

```
//Response before JSON conversion: 
Response {
  [Symbol(realm)]: null,
  [Symbol(state)]: {
    aborted: false,
    rangeRequested: false,
    timingAllowPassed: true,
    requestIncludesCredentials: true,
    type: 'default',
    status: 403,
    timingInfo: {
      startTime: 26454.134393999993,
      redirectStartTime: 0,
      redirectEndTime: 0,
      postRedirectStartTime: 26454.134393999993,
      finalServiceWorkerStartTime: 0,
      finalNetworkResponseStartTime: 0,
      finalNetworkRequestStartTime: 0,
      endTime: 0,
      encodedBodySize: 134,
      decodedBodySize: 134,
      finalConnectionTimingInfo: null
    },
    cacheState: '',
    statusText: 'Forbidden',
    headersList: HeadersList {
      cookies: null,
      [Symbol(headers map)]: [Map],
      [Symbol(headers map sorted)]: null
    },
    urlList: [ [URL] ],
    body: { stream: undefined }
  },
  [Symbol(headers)]: HeadersList {
    cookies: null,
    [Symbol(headers map)]: Map(13) {
      'server' => [Object],
      'date' => [Object],
      'content-type' => [Object],
      'content-length' => [Object],
      'connection' => [Object],
      'x-request-id' => [Object],
      'x-tb-r' => [Object],
      'x-frame-options' => [Object],
      'access-control-allow-origin' => [Object],
      'access-control-allow-headers' => [Object],
      'access-control-allow-methods' => [Object],
      'access-control-max-age' => [Object],
      'strict-transport-security' => [Object]
    },
    [Symbol(headers map sorted)]: null
  }
}
```

```
//Data received from Tinybird: 
undefined
```

**Comment by berkyzl** [comment]
Quick update on the Tinybird API issue I reported: I found that the problem was due to the hardcoded `us-east `region in the Tinybird URL.

For a more flexible solution, we can either use an environment variable for the API region (like TINYBIRD_API_REGION) or utilize the global API endpoint `api.tinybird.co` which doesn't require specifying a region.

I'm closing this issue with this resolution. Thanks for the support.
